File: ralph-workflow/src/guidelines/go.rs
//! Go-specific review guidelines
//!
//! Contains guidelines for Go projects including web frameworks like Gin, Chi, Fiber, and Echo.

use super::base::ReviewGuidelines;
use crate::language_detector::ProjectStack;

/// Add Go-specific guidelines to the review
pub fn add_guidelines(guidelines: ReviewGuidelines, stack: &ProjectStack) -> ReviewGuidelines {
    let base = ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Run go fmt and golint".to_string(),
                "Check all error returns".to_string(),
                "Use defer for cleanup".to_string(),
                "Keep functions short and focused".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Validate input bounds before slice operations".to_string(),
                "Use crypto/rand for security-sensitive random numbers".to_string(),
                "Check for SQL injection in database queries".to_string(),
            ])
            .collect(),
        performance_checks: guidelines
            .performance_checks
            .into_iter()
            .chain([
                "Pre-allocate slices when size is known".to_string(),
                "Use sync.Pool for frequently allocated objects".to_string(),
                "Consider goroutine leaks".to_string(),
            ])
            .collect(),
        testing_checks: guidelines
            .testing_checks
            .into_iter()
            .chain([
                "Use table-driven tests".to_string(),
                "Test error paths explicitly".to_string(),
                "Use testify or similar for assertions".to_string(),
            ])
            .collect(),
        idioms: guidelines
            .idioms
            .into_iter()
            .chain([
                "Accept interfaces, return structs".to_string(),
                "Make the zero value useful".to_string(),
                "Don't communicate by sharing memory".to_string(),
            ])
            .collect(),
        anti_patterns: guidelines
            .anti_patterns
            .into_iter()
            .chain([
                "Don't ignore returned errors".to_string(),
                "Avoid init() when possible".to_string(),
                "Don't use panic for normal error handling".to_string(),
            ])
            .collect(),
        ..guidelines
    };

    if stack
        .frameworks
        .iter()
        .any(|f| matches!(f.as_str(), "Gin" | "Chi" | "Fiber" | "Echo"))
    {
        add_go_web_guidelines(base)
    } else {
        base
    }
}

fn add_go_web_guidelines(guidelines: ReviewGuidelines) -> ReviewGuidelines {
    ReviewGuidelines {
        quality_checks: guidelines
            .quality_checks
            .into_iter()
            .chain([
                "Use proper error handling in handlers".to_string(),
                "Use context for cancellation".to_string(),
                "Structure handlers and middleware properly".to_string(),
            ])
            .collect(),
        security_checks: guidelines
            .security_checks
            .into_iter()
            .chain([
                "Set proper CORS headers".to_string(),
                "Validate input in handlers".to_string(),
            ])
            .collect(),
        ..guidelines
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_go_guidelines() {
        let stack = ProjectStack {
            primary_language: "Go".to_string(),
            secondary_languages: vec![],
            frameworks: vec![],
            has_tests: true,
            test_framework: Some("go test".to_string()),
            package_manager: Some("Go modules".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Go-specific checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("error") || c.contains("golint")));
        assert!(guidelines.anti_patterns.iter().any(|c| c.contains("panic")));
    }

    #[test]
    fn test_go_gin_guidelines() {
        let stack = ProjectStack {
            primary_language: "Go".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Gin".to_string()],
            has_tests: true,
            test_framework: Some("go test".to_string()),
            package_manager: Some("Go modules".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Go web framework checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("handlers") || c.contains("context")));
        assert!(guidelines
            .security_checks
            .iter()
            .any(|c| c.contains("CORS") || c.contains("input")));
    }

    #[test]
    fn test_go_chi_guidelines() {
        let stack = ProjectStack {
            primary_language: "Go".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Chi".to_string()],
            has_tests: true,
            test_framework: Some("go test".to_string()),
            package_manager: Some("Go modules".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Go web framework checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("middleware")));
    }

    #[test]
    fn test_go_fiber_guidelines() {
        let stack = ProjectStack {
            primary_language: "Go".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Fiber".to_string()],
            has_tests: true,
            test_framework: Some("go test".to_string()),
            package_manager: Some("Go modules".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Go web framework checks
        assert!(guidelines
            .security_checks
            .iter()
            .any(|c| c.contains("CORS")));
    }

    #[test]
    fn test_go_echo_guidelines() {
        let stack = ProjectStack {
            primary_language: "Go".to_string(),
            secondary_languages: vec![],
            frameworks: vec!["Echo".to_string()],
            has_tests: true,
            test_framework: Some("go test".to_string()),
            package_manager: Some("Go modules".to_string()),
        };

        let guidelines = add_guidelines(ReviewGuidelines::default(), &stack);

        // Should have Go web framework checks
        assert!(guidelines
            .quality_checks
            .iter()
            .any(|c| c.contains("context")));
    }
}

File: ralph-workflow/ralph/guidelines/go.py
"""Go-specific review guidelines.

Ported from the canonical Rust implementation in
``ralph-workflow/src/guidelines/go.rs`` and adapted for the Python port.
Includes core Go checks plus framework-specific additions for Gin, Chi,
Fiber, and Echo projects.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterable


class GoGuidelines:
    """Review guidelines for Go codebases.

    Args:
        frameworks: Optional framework names used to add stack-specific guidance.
    """

    __slots__ = (
        "anti_patterns",
        "api_design_checks",
        "concurrency_checks",
        "documentation_checks",
        "frameworks",
        "idioms",
        "observability_checks",
        "performance_checks",
        "quality_checks",
        "resource_checks",
        "secrets_checks",
        "security_checks",
        "testing_checks",
    )

    frameworks: tuple[str, ...]
    quality_checks: list[str]
    security_checks: list[str]
    performance_checks: list[str]
    testing_checks: list[str]
    documentation_checks: list[str]
    idioms: list[str]
    anti_patterns: list[str]
    concurrency_checks: list[str]
    resource_checks: list[str]
    observability_checks: list[str]
    secrets_checks: list[str]
    api_design_checks: list[str]

    def __init__(self, frameworks: Iterable[str] = ()) -> None:
        normalized_frameworks = tuple(frameworks)
        self.frameworks = normalized_frameworks
        self.quality_checks = [
            "Run go fmt and golint.",
            "Check all error returns.",
            "Use defer for cleanup.",
            "Keep functions short and focused.",
        ]
        self.security_checks = [
            "Validate input bounds before slice operations.",
            "Use crypto/rand for security-sensitive random numbers.",
            "Check for SQL injection in database queries.",
        ]
        self.performance_checks = [
            "Pre-allocate slices when size is known.",
            "Use sync.Pool for frequently allocated objects.",
            "Consider goroutine leaks.",
        ]
        self.testing_checks = [
            "Use table-driven tests.",
            "Test error paths explicitly.",
            "Use testify or similar for assertions when it improves clarity.",
        ]
        self.documentation_checks = [
            "Document exported types, functions, and packages with Go-style comments.",
            "Explain concurrency expectations, ownership, and side effects in non-obvious code.",
            "Keep README and API examples aligned with actual package behavior.",
        ]
        self.idioms = [
            "Accept interfaces, return structs.",
            "Make the zero value useful.",
            "Don't communicate by sharing memory.",
        ]
        self.anti_patterns = [
            "Don't ignore returned errors.",
            "Avoid init() when possible.",
            "Don't use panic for normal error handling.",
        ]
        self.concurrency_checks = [
            "Use context propagation and cancellation for goroutines and I/O boundaries.",
            "Protect shared state with clear synchronization strategy and avoid race-prone access.",
            "Ensure goroutines terminate cleanly on shutdown, timeout, or caller cancellation.",
        ]
        self.resource_checks = [
            "Close files, response bodies, and other resources on every path.",
            "Use defer near acquisition sites so cleanup is obvious and reliable.",
            "Check cleanup-related errors when they can change user-visible behavior.",
        ]
        self.observability_checks = [
            "Log errors with enough context to diagnose request, job, or goroutine failures.",
            "Avoid noisy logs inside hot paths and tight retry loops.",
            "Expose meaningful metrics or traces for latency-critical and concurrent operations.",
        ]
        self.secrets_checks = [
            "Load secrets from environment or configuration, not source code.",
            "Do not log credentials, tokens, or sensitive request data.",
            "Review default configuration values for accidental secret exposure.",
        ]
        self.api_design_checks = [
            "Keep package APIs small, consistent, and idiomatic for Go consumers.",
            "Return errors as the final result value and make failure modes explicit.",
            "Pass context.Context explicitly at request and I/O boundaries.",
        ]

        for framework in normalized_frameworks:
            self._apply_framework(framework)

    def _apply_framework(self, framework: str) -> None:
        """Add framework-specific guideline extensions."""
        framework_name = framework.casefold()

        if framework_name in {"gin", "chi", "fiber", "echo"}:
            self.quality_checks.extend(
                [
                    "Use proper error handling in handlers.",
                    "Use context for cancellation.",
                    "Structure handlers and middleware properly.",
                ]
            )
            self.security_checks.extend(
                [
                    "Set proper CORS headers.",
                    "Validate input in handlers.",
                ]
            )

    def summary(self) -> str:
        """Return a short human-readable summary."""
        return (
            f"{len(self.quality_checks)} quality checks, "
            f"{len(self.security_checks)} security checks, "
            f"{len(self.anti_patterns)} anti-patterns"
        )

    def total_checks(self) -> int:
        """Return the total number of configured checks."""
        return sum(
            [
                len(self.quality_checks),
                len(self.security_checks),
                len(self.performance_checks),
                len(self.testing_checks),
                len(self.documentation_checks),
                len(self.idioms),
                len(self.anti_patterns),
                len(self.concurrency_checks),
                len(self.resource_checks),
                len(self.observability_checks),
                len(self.secrets_checks),
                len(self.api_design_checks),
            ]
        )


__all__ = ["GoGuidelines"]
